src/wp-includes/blocks.php

```diff
@@ -402,34 +418,20 @@ function register_block_type_from_metadata( $file_or_folder, $args = array() ) {
 	 * instead of reading a JSON file per-block, and then decoding from JSON to PHP.
 	 * Using a static variable ensures that the metadata is only read once per request.
 	 */
-	static $core_blocks_meta;
-	if ( ! $core_blocks_meta ) {
-		$core_blocks_meta = require ABSPATH . WPINC . '/blocks/blocks-json.php';
-	}
 
 	$metadata_file = ( ! str_ends_with( $file_or_folder, 'block.json' ) ) ?
 		trailingslashit( $file_or_folder ) . 'block.json' :
 		$file_or_folder;
 
-	$is_core_block = str_starts_with( $file_or_folder, ABSPATH . WPINC );
-	// If the block is not a core block, the metadata file must exist.
-	$metadata_file_exists = $is_core_block || file_exists( $metadata_file );
-	if ( ! $metadata_file_exists && empty( $args['name'] ) ) {
-		return false;
-	}
-
-	// Try to get metadata from the static cache for core blocks.
-	$metadata = array();
-	if ( $is_core_block ) {
-		$core_block_name = str_replace( ABSPATH . WPINC . '/blocks/', '', $file_or_folder );
-		if ( ! empty( $core_blocks_meta[ $core_block_name ] ) ) {
-			$metadata = $core_blocks_meta[ $core_block_name ];
-		}
-	}
+	$metadata_file_exists = file_exists( $metadata_file );
```

Comment:
@spacedmonkey See https://github.com/WordPress/wordpress-develop/pull/7303#pullrequestreview-2332161245, as this was discussed in that thread.

The `file_exists()` checks here aren't new, they are already happening today. While ideally we can remove them, there is additional risk to doing that, so it's better to leave that for a separate follow-up ticket.

Reply:
I did some benchmarking to see where we're at.

### Methodology

Using my laptop with nginx, php 8.3.12, opcache enabled, no xdebug, I tested on three sets of code. 
* 66afbbdb519d - The trunk commit at the base of this PR as of now.
* db067b0b77 - The latest commit on this PR
* db067b0b77 + custom change - I skipped the `file_exists()` check if getting from a collection of registered metadata.

**The custom change:**
```php
	if ( $registry_metadata ) {
		$metadata = $registry_metadata;
		$metadata_file_exists = true;
	} elseif ( $metadata_file_exists ) {
		$metadata = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
		$metadata_file_exists = file_exists( $metadata_file );
	} else {
		$metadata = array();
	}
```



### Methodology - WooCommerce

I timed the `register_blocks()` function like so:
```php
	public function register_blocks() {
		$this->register_block_metadata();
		$block_types = $this->get_block_types();

		$t0 = hrtime(true);
		$i = 0;
		foreach ( $block_types as $block_type ) {
			$block_type_class = __NAMESPACE__ . '\\BlockTypes\\' . $block_type;

			new $block_type_class( $this->asset_api, $this->asset_data_registry, new IntegrationRegistry() );
		}
		$t1 = hrtime(true);
		$elap = $t1 - $t0;
		$elap = round( $elap / 1e6, 2 );

		$variation = 'woocommerce-yespr';
		$log_file = '/tmp2/elap_log_' . $variation . '.txt';
		file_put_contents($log_file, $elap . "\n", FILE_APPEND);
	}
```

Then requested the twentytwentyfour homepage while logged out 1000 times, and took some basic stats on the recorded numbers.

### Methodology - Core

Most of the blocks are separately registered on init hooks, so I timed init as a whole, there not being an easy way to find a block registration loop.

```php
$t0 = hrtime(true);
do_action( 'init' );
$t1 = hrtime(true);
$elap = $t1 - $t0;
$elap = round( $elap / 1e6, 2 );
$variation = 'init-yespr-skip-file-exists';
$log_file = '/tmp2/elap_log_' . $variation . '.txt';
file_put_contents($log_file, $elap . "\n", FILE_APPEND);
```

Then requested the twentytwentyfour homepage while logged out 1000 times, and took some basic stats on the recorded numbers.

### Results - Core `do_action( 'init' )`

| Scenario | Number of values | Median (p50) (ms) | 75th percentile (p75) (ms) | Standard deviation (ms) |
|----------|------------------|--------------------|-----------------------------|--------------------------|
| Without PR | 1002 | **4.45** | 4.71 | 0.56 |
| With PR | 1003 | **4.69** | 4.92 | 0.49 |
| With PR, skipping file_exists | 1002 | **4.45** | 4.64 | 0.46 |

Commentary: Here, we're not hoping for an improvement, only to make sure we're not regressing. We see a slight regression. If we check the code before this PR, it was indeed skipping the `file_exists()` check for core blocks:

https://github.com/WordPress/wordpress-develop/blob/66afbbdb519d4b21a2ed486de64e32eda3f5adde/src/wp-includes/blocks.php#L414-L416

So here we're re-adding it with a minor penalty.

### Results - WooCommerce `register_blocks()`

| Scenario | Number of values | Median (p50) (ms) | 75th percentile (p75) (ms) | Standard deviation (ms) |
|----------|------------------|--------------------|-----------------------------|--------------------------|
| Without PR | 1002 | **9.96** | 10.36 | 0.61 |
| With PR | 1004 | **8.25** | 8.60 | 0.63 |
| With PR, skipping file_exists | 1003 | **8.04** | 8.37 | 0.57 |

Commentary: We see a good improvement form this PR already, but we can push it even further if we skip the file_exists check. However, we have less guarantee that a third party would have everything set up correctly.

### Conclusions

So it seems we can sneak out a minor edge if we skip the check, and core only is at least skipping this check prior to this PR. I suppose we could either:
* Always assume the file exists if you are using the using the metadata registry
* Or reintroduce the "if you are a core block, we assume the file exists without checking" exception

Reply:
@mreishus Thank you for running these detailed benchmarks.

The differences don't seem significant, however it's a fair point that at least for core blocks the `file_exists` checks were previously not made. And that's reasonable, as we have control over Core and therefore can reliably assume that Core's own files exist.

How about we bring back that `$is_core_block = str_starts_with( $file_or_folder, ABSPATH . WPINC )` variable and use it to avoid the `file_exists()` check for Core blocks like before? It's a bit strange from a code perspective to have this differentiation, but it makes sense from the perspective that we don't want to cause any performance regression here.

That said, I continue to think the topic of potentially bypassing `file_exists()` even for plugins should be discussed via a separate ticket, as that would be a separate improvement with its own potential caveats.

Reply:
Sounds great, I've pushed 6d75a5953cd4acc8112aa0736ca0f3304e3728f9 to give us parity with what we had before.